src/server/server.ml

```diff
@@ -0,0 +1,100 @@
+  Foundation, either version 3 of the License, or (at your option) any later
+  version.
+
+  scilla is distributed in the hope that it will be useful, but WITHOUT ANY
+  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+  A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License along with
+  scilla.  If not, see <http://www.gnu.org/licenses/>.
+*)
+
+open Core
+open DebugMessage
+open ErrorUtils
+open Api
+
+(* You can swap the RPC engine, by using a different monad here,
+   note however that if you are using an asynchronous one, like
+   lwt or async, you should also use their specific IO functions
+   including the print functions.
+
+   You can easily put [ExnM] here and the code would stay unchanged. *)
+module M = Idl.IdM
+module IDL = Idl.Make (M)
+
+module Server = API (IDL.GenServer ())
+
+(* Makes a handler that executes the given [callback] with [args],
+   caches the result using the LRU cache and returns it. **)
+let mk_handler callback args =
+  let open IDL.ErrM in
+  try
+    let result = callback @@ Some args in
+    (* TODO: implement AST caching here *)
+    return result
+  with FatalError msg -> return_err (Idl.DefaultError.InternalError msg)
+
+(* Request handler. *)
+let handler rpc conn =
+  let ic = Unix.in_channel_of_descr conn in
+  let oc = Unix.out_channel_of_descr conn in
+  let msg = Caml.input_line ic in
+  let req = Jsonrpc.call_of_string msg in
+  (* Here we're calling [M.run] to make sure that we are running the process,
+     this is not much of a problem with [IdM] or [ExnM], but in general we
+     should ensure that the computation is started by a runner *)
+  let res = M.run (rpc req) in
+  let str = Jsonrpc.string_of_response res in
+  IPCUtil.send_delimited oc str
+
+(* Listen on the given [sock_path] and process requests.
+   The [num_pending] is the maximal number of pending requests. *)
+let setup ~sock_path ~num_pending =
+  (* Remove any existing socket file *)
+  Unix.(try unlink sock_path with Unix_error (ENOENT, _, _) -> ());
+  (* Ensure that socket directory exists *)
+  Unix.mkdir_p ~perm:0o0755 (Filename.dirname sock_path);
+  let socket = Unix.(socket ~domain:PF_UNIX ~kind:SOCK_STREAM ~protocol:0) in
+  Unix.bind socket ~addr:(Unix.ADDR_UNIX sock_path);
+  Unix.listen socket ~backlog:num_pending;
+  pout @@ Printf.sprintf "Scilla Server is listening on %s\n" sock_path;
+  Out_channel.flush stdout;
+  socket
+
+let rec serve rpc ~socket =
+  let conn, _ = Unix.accept socket in
+  handler rpc conn;
+  Unix.close conn;
+  serve rpc ~socket
+
+let sock_path = "/tmp/scilla-server.sock"
+
+let num_pending = 5
+
+let start ?(sock_path = sock_path) ?(num_pending = num_pending) =
+  pout "Starting Scilla server%s...\n";
+  Out_channel.flush stdout;
+  let runner args =
+    let output, _ = Runner.run args in
```

Comment:
I think we need an error check here, similar to what is being done in `scilla_runner.ml`, and if there's an error, the `error` field in JSON-RPC must be set (instead of the `result` field).

Reply:
In the current implementation the `"error"` field is set in the `mk_handler` function (the `IDL.ErrM` does that implicitly): the `IDL.ErrM.return_err ` "sets" the `"error"` field and the `IDL.ErrM.return` function "sets" the `"result"` field